### ISSUE TITLE
react/no-unused-prop-types cannot find props being used inside helper functions

### ISSUE THREAD

**Issue by renanvalentin** [yannickcr/eslint-plugin-react#1135]
```javascript

const showAlert = props => **props.status** === PENDING;

const Alert = (props) => {
  if (showAlert(props)) {
    return (
      <div />
    );
  }

  return <noscript/>;
};

Alert.propTypes = {
  **status**: PropTypes.oneOf([PENDING, ACCEPTED, CANCELED])
};
```

In this case `props.status` is being used inside of the function `showAlert`.

**Comment by ljharb** [comment]
The proper way to handle this is not to pass the entire props object around. It's not practical or often possible for the plugin to follow references, and the only thing that should get a props object is an element.

In other words, pull `status` out prior to passing it to the function.

**Comment by MethodenMann** [comment]
ok i dont have exactly the same situation. This is my simplified component:

```
import React from 'react'

const MyComp = props => {
  const locales = ['de', 'en']

  const handleFlagClick = locale => {
    props.onLocaleChange(locale)
  }
  return (
    <div>
      {
        locales.map(locale => (
          <img
            key={locale}
            onClick={() => handleFlagClick(locale)}
          />
        ))
      }
    </div>
  )
}

MyComp.propTypes = {
  onLocaleChange: React.PropTypes.func
}

export default MyComp

```

And i get the following linting error:
t

>   24:19  error  'onLocaleChange' PropType is defined but prop is never used  react/no-unused-prop-types


I figured out its because of the map. It this by design?

**Comment by jseminck** [comment]
Hello,

I ran your latest test case in a PR where we fixed the following issue: https://github.com/yannickcr/eslint-plugin-react/issues/1183 and the test case passed. Prior to the changes from that specific PR, the test case was failing.

Removing the `.map()` code in the `render()` would make the component valid under the current implementation.

So I am quite confident those changes are also fixing the issue that you see here. And yes I would think that this to be a bug. Do others agree?
